src/terrain/KM_Terrain.pas

```diff
@@ -585,31 +622,210 @@ function TKMTerrain.ScriptTryObjectSet(X, Y: Integer; aObject: Byte): Boolean;
+        end else
+          SetErrorNSetResult(tctHeight, HasErrorOnTile, ErrorTypesOnTile, Result);
+      end;
+
+      //Update object if needed
+      if tctObject in T.ChangeSet then
+      begin
+        if TrySetTileObject(T.X, T.Y, T.Obj, DiagChanged, False) then
+        begin
+          UpdateRect(Rect, T.X, T.Y);
+          DiagonalChangedTotal := DiagonalChangedTotal or DiagChanged;
+        end else
+          SetErrorNSetResult(tctObject, HasErrorOnTile, ErrorTypesOnTile, Result);
+      end;
+    end else
+    begin
+      HasErrorOnTile := True;
+      //When tile is out of map coordinates we treat it as all operations failure
+      if tctTerrain in T.ChangeSet then
+        Include(ErrorTypesOnTile, tctTerrain);
+      if tctHeight in T.ChangeSet then
+        Include(ErrorTypesOnTile, tctHeight);
+      if tctObject in T.ChangeSet then
+        Include(ErrorTypesOnTile, tctObject);
+    end;
+
+    // Save error info, if there was some error
+    if HasErrorOnTile then
+    begin
+      if Length(aErrors) = ErrCnt then
+        SetLength(aErrors, ErrCnt + 16);
+      aErrors[ErrCnt].X := T.X;
+      aErrors[ErrCnt].Y := T.Y;
+      aErrors[ErrCnt].ErrorsIn := ErrorTypesOnTile;
+      Inc(ErrCnt);
+    end;
+
+    if not Result and aRevertOnFail then
+      Break;
+  end;
+
+  if not Result and aRevertOnFail then
+  begin
+    //Restore backup Land, when revert needed
+    for I := 1 to fMapY do
+      for J := 1 to fMapX do
+        Land[I, J] := BackupLand[I-1][J-1];
+    SetLength(BackupLand, 0); // Release dynamic array memory. This array can be huge, so we should clear it as fast as possible
+  end
+  else
+  begin
+    if IsRectInit(HeightRect) then
+      gTerrain.UpdateLighting(KMRectGrow(HeightRect, 2)); // Update Light only when height was changed
+
+    gTerrain.UpdatePassability(KMRectGrowTopLeft(Rect));
+    gTerrain.UpdateWalkConnect([wcWalk, wcRoad, wcFish, wcWork], KMRectGrowTopLeft(Rect), DiagonalChangedTotal);
+  end;
+
+  //Cut errors array to actual size
+  if Length(aErrors) <> ErrCnt then
+    SetLength(aErrors, ErrCnt);
+end;
+
+
+// Try to set an tile (Terrain and Rotation) from the script. Failure is an option
+function TKMTerrain.ScriptTrySetTile(X, Y: Integer; aType, aRot: Byte): Boolean;
+begin
+  Result := False;
+  if not TileInMapCoords(X, Y) then
+    Exit;
+  Result := TrySetTile(X, Y, aType, aRot);
+end;
+
+
+// Try to set an tile Height from the script. Failure is an option
+function TKMTerrain.ScriptTrySetTileHeight(X, Y: Integer; aHeight: Byte): Boolean;
+begin
+  Result := False;
+  if not TileInMapCoords(X, Y) then
+    Exit;
+  Result := TrySetTileHeight(X, Y, aHeight);
+end;
+
+
+// Try to set an object from the script. Failure is an option
+function TKMTerrain.ScriptTrySetTileObject(X, Y: Integer; aObject: Byte): Boolean;
+begin
+  Result := False;
+  if not TileInMapCoords(X, Y) then
+    Exit;
+  Result := TrySetTileObject(X, Y, aObject);

```

Comment:
One line could be better instead:
Result := TileInMapCoords(X, Y) and TrySetTileObject(X, Y, aObject);